### PR TITLE
refactor: validate config

### PR DESCRIPTION
**Pull request by FGYFFFF**
#### What type of PR is this?

refactor
#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.
- [x] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) Translate the PR title into Chinese.
重构参数校验的配置行为

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en: refactor the config for validating
zh(optional): 重构参数校验的配置行为

### CODE REVIEW

**Review by li-jin-gou — DISMISSED** [review]

File: pkg/app/server/binding/default.go [hunk 1]
```diff
@@ -374,36 +374,53 @@ func (b *defaultBinder) bindNonStruct(req *protocol.Request, v interface{}) (err
 var _ StructValidator = (*defaultValidator)(nil)
 
 type defaultValidator struct {
-	once     sync.Once
 	validate *validator.Validator
 }
 
 func NewDefaultValidator(config *ValidateConfig) StructValidator {
-	return &defaultValidator{}
+	vd := validator.New("vd").SetErrorFactory(defaultErrorFactory)
```
Comment (welkeyever) [inline pkg/app/server/binding/default.go:381]: 这个"vd"可以塞到config里面让用户能够自定义吗？从config传进来

File: pkg/app/server/binding/default.go [hunk 2]
```diff
@@ -374,36 +374,53 @@ func (b *defaultBinder) bindNonStruct(req *protocol.Request, v interface{}) (err
...
+	if config != nil && config.ErrFactory != nil {
+		vd.SetErrorFactory(config.ErrFactory)
+	}
+	return &defaultValidator{
+		validate: vd,
+	}
+}
+
+// Error validate error
+type defaultError struct {
```
Comment (welkeyever) [inline pkg/app/server/binding/default.go:391]: 这个是一个公共结构体？不需要搞个defaul前缀的名字吧？validateError 是不是好些？

File: pkg/app/server/binding/default.go [hunk 3]
```diff
@@ -374,36 +374,53 @@ func (b *defaultBinder) bindNonStruct(req *protocol.Request, v interface{}) (err
...
+	FailPath, Msg string
+}
+
+// Error implements error interface.
+func (e *defaultError) Error() string {
+	if e.Msg != "" {
```
Comment (welkeyever) [inline pkg/app/server/binding/default.go:397]: 这个地方有点怪，e.Msg 为空的时候为啥还要搞个错误返回？

File: pkg/route/engine.go [hunk 4]
```diff
@@ -558,13 +558,21 @@ func (engine *Engine) ServeStream(ctx context.Context, conn network.StreamConn)
 
 func (engine *Engine) initBinderAndValidator(opt *config.Options) {
 	// init validator
-	engine.validator = binding.DefaultValidator()
 	if opt.CustomValidator != nil {
 		customValidator, ok := opt.CustomValidator.(binding.StructValidator)
 		if !ok {
 			panic("customized validator can not implement binding.StructValidator")
```
Comment (welkeyever) [inline pkg/route/engine.go:564]: "customized validator can not implement binding.StructValidator" -> "customized validator does not implement binding.StructValidator"

File: pkg/route/engine.go [hunk 5]
```diff
@@ -558,13 +558,21 @@ func (engine *Engine) ServeStream(ctx context.Context, conn network.StreamConn)
...
 		}
 		engine.validator = customValidator
+	} else {
+		engine.validator = binding.NewDefaultValidator(binding.NewValidateConfig())
+		if opt.ValidateConfig != nil {
+			vConf, ok := opt.ValidateConfig.(*binding.ValidateConfig)
+			if !ok {
+				panic("validate config error")
```
Comment (welkeyever) [inline pkg/route/engine.go:572]: 报错信息再详细点呢

File: pkg/app/server/binding/default.go [hunk 6]
```diff
@@ -185,14 +185,17 @@ func (b *defaultBinder) bindTag(req *protocol.Request, v interface{}, params par
 		decoder := cached.(decoderInfo)
 		return decoder.decoder(req, params, rv.Elem())
 	}
-
+	validateTag := "vd"
```
Comment (welkeyever) [inline pkg/app/server/binding/default.go:188]: 这个抽成`defaultValidateTag`常量放到最上面去？

File: pkg/app/server/binding/default.go [hunk 7]
```diff
@@ -374,34 +381,40 @@ func (b *defaultBinder) bindNonStruct(req *protocol.Request, v interface{}) (err
...
 
 type defaultValidator struct {
-	validate *validator.Validator
+	validateTag string
+	validate    *validator.Validator
 }
 
 func NewDefaultValidator(config *ValidateConfig) StructValidator {
-	vd := validator.New("vd").SetErrorFactory(defaultErrorFactory)
+	validateTag := "vd"
```
Comment (welkeyever) [inline pkg/app/server/binding/default.go:389]: 用刚刚抽出来那个常量？

File: pkg/app/server/binding/default.go [hunk 8]
```diff
@@ -374,36 +381,63 @@ func (b *defaultBinder) bindNonStruct(req *protocol.Request, v interface{}) (err
 var _ StructValidator = (*defaultValidator)(nil)
 
 type defaultValidator struct {
```
Comment (welkeyever) [inline pkg/app/server/binding/default.go:383]: 这个结构体是不是也不需要以`default`开头？

Comment (FGYFFFF) [reply pkg/app/server/binding/default.go:383]: 可以的。我改下

File: pkg/app/server/binding/default.go (hunk 6 above)
Comment (FGYFFFF) [reply pkg/app/server/binding/default.go]: done

File: pkg/app/server/binding/default.go (hunk 7 above)
Comment (FGYFFFF) [reply pkg/app/server/binding/default.go]: done

File: pkg/app/server/binding/default.go (hunk 8 above)
Comment (FGYFFFF) [reply pkg/app/server/binding/default.go]: done

**Review by Duslia — APPROVED** [review]

**Review by welkeyever — APPROVED** [review]
LGTM

**Review by li-jin-gou — APPROVED** [review]
LGTM

**Pull request merged**